### ISSUE TITLE
Export FMU with Bus Inport/Outport

### ISSUE THREAD

**Issue by joaha058** [CATIA-Systems/FMIKit-Simulink#72]
Hi!

I'm evaluating the FMIKit to see if we can use it in our project to exchange models between simulink and Dymola. 

Tried version 2.6 and the GRT export.  It seems like it is not possible to import the generated FMU into Simulink again if the interface is of type Bus.  It looks like there are some problems with the modelDescription.xml generation (Missing attributes or something)

Our Simulink Models have bus interfaces with a couple of hundreds signals in it and it wouldn't be possible to change the interface to scalars/vectors.

Is this a known problem with FMIKit? Or is it a problem with the FMI standard which only allows scalar/vectors in the interface?

Regards
Joakim

**Comment by kojojo1** [comment]
Hi,
I have observed same issue. It is significant limitation of the FMI-Kit because almost every bigger model use a bus.
I have some FMUs created by Modelica and it seams that Modelica can export FMUs with a bus.
See the name attribute in ModelDescription.xml example from Modelica:

`<ScalarVariable description="Thermal conductivity" name="medium.lamda" variability="fixed" causality="parameter" valueReference="16777219">`
`<ScalarVariable description="Kinematic viscosity" name="medium.nue" variability="fixed" causality="parameter" valueReference="16777220">
`

the 'bus' is described by dot notation.

**Comment by chrbertsch** [comment]
It would be very helpful to map the "dot"-notation from the modeldescription.xml to a Simulink bus, as e.g. the FMI import by Mathworks can: https://de.mathworks.com/help/simulink/slref/using-bus-signals-and-structure-parameters-in-the-fmu-import-block.html

**Comment by t-sommer** [comment]
@joaha058, @kojojo1, @chrbertsch, can you please share a model or Simulink built-in example that reproduces the problem?

**Comment by kojojo1** [comment]
There is the example. The ModelDescription.xml is empty after compilation. I've used VS2012 Matlab2016, FMI Kit 2.6.0   
[busIssue.zip](https://github.com/CATIA-Systems/FMIKit-Simulink/files/3635454/busIssue.zip)
The buses definition are attached in initBuses.mat file .

**Comment by t-sommer** [comment]
So what you're looking for is the XML like below?

```xml
<ScalarVariable name="InputBus.InputCont" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="InputBus.InputDiscrete" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="OutBus.OutputCont" causality="output">
  <Real/>
</ScalarVariable>
<ScalarVariable name="OutBus.OutputDiscrete" causality="output">
  <Real/>
</ScalarVariable>
```

**Comment by kojojo1** [comment]
yes, exactly. Of course the Bus can be more complex. Look at the attachment. 

[busIssueSub.zip](https://github.com/CATIA-Systems/FMIKit-Simulink/files/3640933/busIssueSub.zip)

I would expect following ModelDescription.xml in this case:

`<ScalarVariable name="InputBus.InputCont" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="InputBus.InputDiscrete" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="InputBus.Struct.ChannelA" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="InputBus.Struct.SubBus.ChannelB" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="InputBus.Struct.SubBus.ChannelC" causality="input">
  <Real/>
</ScalarVariable>
<ScalarVariable name="OutBus.OutputCont" causality="output">
  <Real/>
</ScalarVariable>
<ScalarVariable name="OutBus.OutputDiscrete" causality="output">
  <Real/>
</ScalarVariable>` 

By the way, I use FMI kit for model generating for National Instruments Veristand. Since Veristand 2019  FMU import is natively supported. I think it has great potential, because the FMI kit gives possibility to generate the model for Veristand without Veristand  :-)

**Comment by meek1** [comment]
Agree, bus support would be extremely useful.

**Comment by t-sommer** [comment]
With the experimental modifications in 23d71816f3636072483fb48b6a5850ee97aeeecd I can export busIssueSub.slx with R2019b. I get the following XML:

```xml
<!-- Inputs -->
<ScalarVariable name="InputBus_h.Struct.SubBus.ChannelB" valueReference="4" causality="input">
  <Real start="0"/>
</ScalarVariable>
<ScalarVariable name="InputBus_h.Struct.SubBus.ChannelC" valueReference="5" causality="input">
  <Real start="0"/>
</ScalarVariable>
<ScalarVariable name="InputBus_h.Struct.ChannelA" valueReference="6" causality="input">
  <Real start="0"/>
</ScalarVariable>
<ScalarVariable name="InputBus_h.InputCont" valueReference="7" causality="input">
  <Real start="0"/>
</ScalarVariable>
<ScalarVariable name="InputBus_h.InputDiscrete" valueReference="8" causality="input">
  <Real start="0"/>
</ScalarVariable>

<!-- Outputs -->
<ScalarVariable name="OutBus.OutputCont" valueReference="9" causality="output">
  <Real start="0"/>
</ScalarVariable>
<ScalarVariable name="OutBus.OutputDiscrete" valueReference="10" causality="output">
  <Real start="0"/>
</ScalarVariable>
```

**Comment by kojojo1** [comment]
That is great! How can I try it?

**Comment by chrbertsch** [comment]
Will this be included in FMI-Kit 2.7?

**Comment by t-sommer** [comment]
> Will this be included in FMI-Kit 2.7?

Yes. It's already included in 2.7 Alpha 2.

**Comment by t-sommer** [comment]
> Is there a summary of new features and bugfixes of FMI 2.7 alpha 2, so that one knows, what to test?

We'll update the changelog before the release. For now you can take a look at the commit messages for a rough overview.
